File: 백준/Gold/2056. 작업/작업.py
# 작업
import sys
input = sys.stdin.readline

N = int(input())
front = [0] * N
time = [0] * N
res = [0] * N
tail = [[] for _ in range(N)]
stack = []

for i in range(N):
    tmp = list(map(int,input().split()))

    time[i] = tmp[0]
    front[i] = tmp[1]

    for j in range(tmp[1]):
        tail[tmp[2+j]-1].append(i)

    if tmp[1] == 0:
        stack.append((i, tmp[0]))
        res[i] = tmp[0]

while stack:

    cur, ct = stack.pop()

    for j in (tail[cur]):
        front[j] -= 1

        res[j] = max(res[j], time[j] + ct)
        
        if front[j] == 0:
            stack.append((j, res[j]))

print(max(res))

File: 백준/Gold/2056. 작업/작업.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        int N = Integer.parseInt(br.readLine());
        int[] time = new int[N];
        int[] res = new int[N];
        int[] front = new int[N];

        ArrayList<ArrayList<Integer>> tail = new ArrayList();
        for (int i = 0; i < N; i++) {
            tail.add(new ArrayList());
        }

        ArrayDeque<int[]> stack = new ArrayDeque<>();

        int max = 0;

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());

            time[i] = Integer.parseInt(st.nextToken());
            front[i] = Integer.parseInt(st.nextToken());

            if (front[i] == 0) {
                stack.push(new int[]{i, time[i]});
                continue;
            }

            while (st.hasMoreTokens()) {
                int f = Integer.parseInt(st.nextToken());
                tail.get(f-1).add(i);
            }
        }


        while (!stack.isEmpty()) {
            int[] tmp = stack.pop();

            int cur = tmp[0];
            int ct = tmp[1];

            max = Math.max(max, ct);

            for (int i : tail.get(cur)) {
                front[i]--;

                res[i] = Math.max(res[i], time[i] + ct);

                if (front[i] == 0) {
                    stack.push(new int[]{i, res[i]});
                }
            }
        }

        System.out.println(max);
    }
}
